### ISSUE TITLE
ISSUE #34 New Sanity Check: duplicated properties, 4 items

### ISSUE THREAD

**Issue by Ricarius** [net-lisias-ksp/TweakScale#99]
I am experiencing ' ISSUE #34 New Sanity Check: duplicated properties ' FATAL error warning for 4 items on the most recent Tweakscale pre-release build (2.4.3.11):
S2Structual from AirplanesPlus mod. (This, in my opinion, is a NON-CRITICAL part, as it is not particularly unique and other parts serve a similar purpose. Fix at your leisure.)
awacsRadar from BDArmory (This, in my opinion, is a CRITICAL part, as it is unique and serves a specific purpose. If it is true that that the parts are not the culprits, but the screaming victims, please make this one stop screaming before any others.)
NMB117SEgnine from Nice MKSeries Body (This, in my opinion, is a NON-CRITICAL part, as it is not particularly unique and other parts serve a similar purpose. Fix at your leisure.)
TypeFlankerIntake from Nice MKSeries Body (This, in my opinion, is a NON-CRITICAL part, as it is not particularly unique and other parts serve a similar purpose. Fix at your leisure.)

**Comment by Lisias** [comment]
Hi.

The way to fix the problem is detecting who is double patching the parts and fix that patch.

TweakScale did its job, it detected a no conformity, and deactivated support for that parts to prevent damages. Since it's impossible to know every single part in existence (not to mention new parts being created), the only sensible line of action for TweakScale is to complain about the no-conformity and prevent disasters by deactivating them.

Publish your KSP.log here and we can try to find the problems - it's the only way to fix your instalment, as TweakScale has done its job properly and this mechanism will not be changed.

**Comment by Ricarius** [comment]
Here is the KSP.log file, since I was not able to upload it here due to exceeding the 10MB file size limit:
https://drive.google.com/open?id=1RNZxSnZTWcToR1vQq4gcZmI0KFQGjP4q

**Comment by Lisias** [comment]
Hint: Install ZeroMiniAVC (or delete all MiniAVC.dll after installing something). MiniAVC is causing a lot of trouble on KSP 1.8 and 1.9 , and the maintainer decided to deprecate it.

**Comment by Lisias** [comment]
Got. It's TMasterson5TweakscalePatches . :(

This set of patches are terribly outdated, and I can't fix them myself due the restrictive licensing terms.

More on the matter here: https://forum.kerbalspaceprogram.com/index.php?/topic/179030-ksp-141-tweakscale-under-lisias-management-24311-2020-0303/&do=findComment&comment=3658877

Link for the TMasterson5's : https://forum.kerbalspaceprogram.com/index.php?/topic/181010-tmasterson5-mod-family/

These patches are pretty popular, and it gave me some work last year. I would gladly fix them and apply a pull request for the maintainer if he had published the Add'On on github or something.

But being CC-BY-NC-ND, I cannot fix it and redistributed myself, and without a github/gitlab/bitbucket or something, I can't apply the fixes. I have my hands tied.

It's a pitty that TMasterso5 is not logging on the Forum since last year too so we can't even ask for a less restrictive licensing terms, or at least sent him the fixes so he can publish a new version.

Sorry, I can't help further.

**Comment by Lisias** [comment]
You can fix your patches yourself by doing that follows:

1) Edit every `type = ` on a TweakScale patch to `%type =` (note the percent sign).
2) Add `:NEEDS[TweakScale]` on every `%MODULE[TweakScale]`.

The ending results must be something like:

```
     %MODULE[TweakScale]:NEEDS[TweakScale]
     {
           <something>
           %type = <something>
     }
```

**Comment by Ricarius** [comment]
The screaming has stopped, thank you a lot.

**Comment by Ricarius** [comment]
After using CKAN to auto-update many of my addons after an extended period of time away from KSP, I am again experiencing ' ISSUE #34 New Sanity Check: duplicated properties ' FATAL error warning for 14 items on the most recent Tweakscale release build (2.4.3.21):
- smallwingConnector1
- smallwingConnector2
- smallwingConnector3
- smallwingConnector4
- smallwingConnector5
- smallwingConnector6
- smallwingConnectortip
- waterjet
- smallCtrlSrf
- smallRadialEngine
- smallRadialEngine.v2
- smallHardpoint
- mediumDishAntenna
- smallClaw

[KSP.log](https://github.com/net-lisias-ksp/TweakScale/files/5561974/KSP.log)

**Comment by Lisias** [comment]
> After using CKAN to auto-update many of my addons after an extended period of time away from KSP, I am again experiencing ' ISSUE #34 New Sanity Check: duplicated properties ' FATAL error warning for 14 items on the most recent Tweakscale release build (2.4.3.21):

Yep. It's unfortunate that TweakScale still have to handle these problems even nowadays, but since so many good and loved add'ons are unmaintained nowadays, it end up that TweakScale is the one in the position of detecting and fixing these problems.

Checking the log for one of the parts mentioned (and thanks for the detailed log!), I found:

```
[LOG 13:27:56.807] Config(PART) AirplanePlus/Parts/Aero/smallwings/halfwing/smallwingConnector1
[LOG 13:21:18.847] Applying update BDArmory/MMPatches/000000_HitpointModule/@PART[*] to AirplanePlus/Parts/Aero/smallwings/halfwing.cfg/PART[smallwingConnector1]
[LOG 13:21:21.734] Applying update LShipParts/LShipPartsRequired/LBP_tweakableComplete/@PART[small*]:NEEDS[TweakScale] to AirplanePlus/Parts/Aero/smallwings/halfwing.cfg/PART[smallwingConnector1]
[LOG 13:21:40.480] Applying update AirplanePlusThings/APP_TweakScale/@PART:HAS[#manufacturer[Kerbal?Standard],#category[Aero],#bulkheadProfiles[srf]]:NEEDS[Tweakscale]:AFTER[AirplanePlus] to AirplanePlus/Parts/Aero/smallwings/halfwing.cfg/PART[smallwingConnector1]
[LOG 13:28:01.770] PartLoader: Compiling Part 'AirplanePlus/Parts/Aero/smallwings/halfwing/smallwingConnector1'
[LOG 13:28:01.780] PartLoader: Part 'AirplanePlus/Parts/Aero/smallwings/halfwing/smallwingConnector1' has no database record. Creating.
[LOG 13:28:01.781] DragCubeSystem: Creating drag cubes for part 'smallwingConnector1'
[LOG 13:30:32.621] [TweakScale] WARNING: **FATAL** Found a showstopper problem on smallwingConnector1 (Wing Connector Type F).
[LOG 13:30:32.621] [TweakScale] ERROR: **FATAL** Part smallwingConnector1 (Wing Connector Type F) has a fatal problem due having duplicated properties - see issue [#34]( https://github.com/net-lisias-ksp/TweakScale/issues/34 ). at error:0
```

Unfortunately, AirplanePlusThings are blindly double patching things, bluntly ignoring that the part was already patched.

My best advice is to replace AirplanePlusThigs with [TweakScale Companion for Airplane Plus](https://github.com/net-lisias-ksp/TweakScaleCompanion_APP). TSC_APP does things right, and if by any reason something borks on it, open an issue there and I will fix it. :)

Cheers.

**Comment by Lisias** [comment]
Humm... wait... not all parts affected are from AirplanePlus...

Please install also [TweakScale Companion for SMCE](https://github.com/net-lisias-ksp/TweakScaleCompanion_SMCE) . It will fix some problems related to Large Ship Parts. It's unfortunate that SMCE had quit the Scene, and so his add'ons are currently unmaintained - but I'm doing my best to keep them working. These are magnificent add'ons, I wish we could lure SMCE back. :)

**Comment by Lisias** [comment]
let me know if these two TSC fixes your problem, so I can close this!

Cheers!

**Comment by Ricarius** [comment]
Surprisingly, it seems fixing my load order solved the problem. Airplane Plus Things appears to be working fine for the parts now, I just had to install Tweakscale before any other mods. I am still getting 14 sanity checks, but they are non-fatal now, and not quite a concern for me anymore. I could try switching over to the addons you recommended if you believe that might get rid of the non-fatal sanity checks, but I'm hesitant because I don't want to risk the fatal ones returning as a result of further tinkering.

**Comment by Lisias** [comment]
> Surprisingly, it seems fixing my load order solved the problem. Airplane Plus Things appears to be working reasonably okay for the parts now, I just had to install Tweakscale before any other mods.

Extremely surprising! The add'on installation order **mades no difference**, as the load order of the add'ons is determinated by the MM directives or, in their absence, by more os less alfanumerical order (more or less because the path are escaped to URLs, and some chars are replaced on the process).

I'm checking the KSP.log.

**Comment by Lisias** [comment]
Found the difference. You didn't installed LShipParts . Without it, the AiplanePlus patches are not being doubled. You would the same result by not installing AiplanePlus Things - with the difference that now you don't have the LShipParts available.

**Comment by Lisias** [comment]
Hum.... On a second look on the first KSP.log, I realised **why** LShipParts was hitting some AirplanePlus parts...

```
[LOG 13:21:21.734] Applying update LShipParts/LShipPartsRequired/LBP_tweakableComplete/@PART[small*]:NEEDS[TweakScale] to AirplanePlus/Parts/Aero/smallwings/halfwing.cfg/PART[smallwingConnector1]
```

It's using a pretty broad wildcard while patching, and it ended up hitting AirplanePlus by accident.

That's the reason why TweakScale Companion for AirplanePlus clean up things before applying its patches.

Since you choose to **do not** install LShipParrs, you will be ok for while. But if you ever install it again, you will need to download the two TweakScale Companions I mentioned before, being them:

* [TweakScale Companion for Airplane Plus](https://github.com/net-lisias-ksp/TweakScaleCompanion_APP)
* [ TweakScale Companion for SMCE](https://github.com/net-lisias-ksp/TweakScaleCompanion_SMCE)

**Comment by Ricarius** [comment]
Interesting! I might just switch over to the two TSComp add-ons just in case I run into any other mods like LShipParts which do the broad sweep for tweakscale compatibility. Hoping that AirplanePlusThings hasn't already done too much damage to the other files if I swap it out.

**Comment by Lisias** [comment]
TweakScale 2.5 beta, right now, is able to "convert" between different scaling rules using a rule of three - so any difference between Thing and the Companion will be seamless converted by TS. (it had it on KSP 1.3 times, but KSP 1.4 changed something very subtle inside, and I only realised what this year...)

Right now, the Beta is safe to be used (just **backup everything** on updates, that thing is beta for a reason) - the only mishap is about scaling parts with variants where the attach points changes, and the only parts doing this now are the Mastodon engine and the Tubes from Making History. Other than having to fix the part position by hand, nothing bad happens (but boy, that is annoying... :P ).

The nice thing about TweakScale 2.5 is that all patches are being meticulously audited and fixed, and rewritten to allow better standards on patching. The not so nice thing about 2.5 Beta is that now and then I let something weird pass trough, and that can be annoying to fix.

**Comment by Ricarius** [comment]
Hmm, replacing Things with the APP Companion resulted in an increase from 14 to 15 non-fatal sanity checks. Not sure if this has to do with the fact that I didn't include the SMCE Companion, since I currently don't have LShipParts included in my modpack I figured I'd exclude it.
[KSP.log](https://github.com/net-lisias-ksp/TweakScale/files/5569563/KSP.log)

**Comment by Ricarius** [comment]
Either way, the fatal sanity checks are gone now, so I think the issue is dealt with once again. Hopefully when 2.5 releases the issue with the non-fatal sanity checks won't be sticking around either.

**Comment by Lisias** [comment]
> Hmm, replacing Things with the APP Companion resulted in an increase from 14 to 15 non-fatal sanity checks.

From that 15,

* 9 are related to Part with Variants that change Mass or Cost. There ones are fixed on TweakScale 2.5 Beta, and will be backported to 2.4.4.0 as soon as I manage to work out the last problem related with variants (attachment points).
* 6 are related to FSBuouyance , that is not supported at all on TS 2.4.3 . It's fixed on 2.5 Beta (as long you use [TweakScale Companion for Firespitter](https://github.com/net-lisias-ksp/TweakScaleCompanion_FS)), and will also be available on TS 2.4.4.0. 

So, yeah. All these Non Fatals are already fixed on the development branch, and will be for sure available on the next Release of TweakScale.
